spec/services/submission/trait_score_parser_spec.rb

```diff
@@ -5,101 +5,234 @@
+        input_is 'id,pa,oo,pl'
+        subject.send(:parse_header)
+        expect(upload.errors[:file]).
+          to eq ['No correct header provided. Please provide the \"Plant accession\" column.']
+      end
+
+      it 'complains if there is neither Plant line nor Plant variety column' do
+        input_is 'id,Plant accession,oo,pl'
+        subject.send(:parse_header)
+        expect(upload.errors[:file]).
+          to eq ['No correct header provided. Please provide either the \"Plant line\" or the \"Plant variety\" column.']
+      end
     end
 
     context 'provided with trait-rich submission' do
       it 'maps columns by name' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['trait 1', 'trait 2'])
-        input_is "id,trait 2,trait 1"
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,trait 2,trait 1"
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 1, 1 => 0})
       end
 
       it 'recognizes names with commas and surrounding white chars' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['trait,1', 'trait 2'])
-        input_is "id,trait 2  ,\"trait,1\""
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,trait 2  ,\"trait,1\""
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 1, 1 => 0})
       end
 
       it 'honors proper trait sorting by index' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['Ctrait', 'Atrait', 'Btrait'])
-        input_is "id,Btrait,Atrait,Ctrait"
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,Btrait,Atrait,Ctrait"
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 2, 1 => 1, 2 => 0})
       end
 
       it 'uses natural ordering when no by-name mapping found' do
         upload.submission.content.update(:step02, trait_descriptor_list: ['Atrait', 'Btrait'])
-        input_is "id,trait 1,trait 2"
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,trait 1,trait 2"
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 0, 1 => 1})
       end
 
       it 'works regardless traits are old or new' do
         td = create(:trait_descriptor, trait: create(:trait, name: 'old trait'))
         upload.submission.content.update(:step02, trait_descriptor_list: [td.id, 'new trait'])
-        input_is "id,new trait,old trait"
-        subject.send(:map_headers_to_traits)
+        input_is "id,Plant accession,oo,Plant line,new trait,old trait"
+        subject.send(:parse_header)
         expect(subject.trait_mapping).
           to eq({0 => 1, 1 => 0})
       end
 
       it 'reports error on repetitive mapping' do
```

Comment:
This is trickier. When you read my previous comment, you'll know there are two strategies to that. And in this case the user is somehow mixing them together - s/he puts some completely different named trait in place, where Atrait scores should be given, yet s/he puts scores for Atrait elsewhere. This results in ambiguity. We should ask the user to follow one approach: either column headers with exact trait names, or consistent column ordering.

I try to explain it to the user in the manual, and the template should be correct anyway, unless the user shuffles column headers for some reason.